File: fsharp/space-age/SpaceAge.fs
module SpaceAge

[<Measure>]
type earthYear

[<Measure>]
type second

type Planet =
    | Mercury
    | Venus
    | Earth
    | Mars
    | Jupiter
    | Saturn
    | Uranus
    | Neptune

let age (planet: Planet) (seconds: int64) : float =
    planet
    |> function
    | Mercury -> 0.2408467<earthYear>
    | Venus -> 0.61519726<earthYear>
    | Earth -> 1.0<earthYear>
    | Mars -> 1.8808158<earthYear>
    | Jupiter -> 11.862615<earthYear>
    | Saturn -> 29.447498<earthYear>
    | Uranus -> 84.016846<earthYear>
    | Neptune -> 164.79132<earthYear>
    |> (*) 31557600.0<second/earthYear>
    |> (/) ((float seconds) * 1.0<second>)

File: clojure/space-age/src/space_age.clj
(ns space-age)

(defn- on-planet [planet-seconds seconds]
  (/ seconds planet-seconds))

(def on-mercury (partial on-planet 7600543.81992))

(def on-venus (partial on-planet 1.9414149052176E7))

(def on-earth (partial on-planet 31557600))

(def on-mars (partial on-planet 5.9354032690079994E7))

(def on-jupiter (partial on-planet 3.74355659124E8))

(def on-saturn (partial on-planet 9.292923628848E8))

(def on-neptune (partial on-planet 5.200418560032001E9))

(def on-uranus (partial on-planet 2.6513700193296E9))
